File: sdk/runanywhere-swift/Sources/RunAnywhere/Data/Network/Services/AuthenticationService.swift
import Foundation

/// Service responsible for authentication and token management
public actor AuthenticationService {

    // MARK: - Properties

    private let apiClient: APIClient
    private var accessToken: String?
    private var refreshToken: String?
    private var tokenExpiresAt: Date?
    private var deviceId: String?
    private var userId: String?  // Optional since API can return null for now if org_level
    private var organizationId: String?
    private let logger = SDKLogger(category: "AuthenticationService")

    // MARK: - Initialization

    public init(apiClient: APIClient) {
        self.apiClient = apiClient
    }

    /// Create and configure authentication services for production/staging
    /// - Parameters:
    ///   - baseURL: The API base URL
    ///   - apiKey: The API key for authentication
    /// - Returns: Tuple of configured (APIClient, AuthenticationService)
    /// - Throws: If authentication fails
    public static func createAndAuthenticate(
        baseURL: URL,
        apiKey: String
    ) async throws -> (apiClient: APIClient, authService: AuthenticationService) {
        let apiClient = APIClient(baseURL: baseURL, apiKey: apiKey)
        let authService = AuthenticationService(apiClient: apiClient)
        await apiClient.setAuthenticationService(authService)

        // Authenticate with backend
        _ = try await authService.authenticate(apiKey: apiKey)

        return (apiClient, authService)
    }

    // MARK: - Public Methods

    /// Authenticate with the backend and obtain access token
    public func authenticate(apiKey: String) async throws -> AuthenticationResponse {
        let deviceId = DeviceIdentity.persistentUUID

        let request = AuthenticationRequest(
            apiKey: apiKey,
            deviceId: deviceId,
            platform: SDKConstants.platform,
            sdkVersion: SDKConstants.version
        )

        logger.debug("Authenticating with backend")

        // Use APIClient for the authentication request (doesn't require auth)
        let authResponse: AuthenticationResponse = try await apiClient.post(
            .authenticate,
            request,
            requiresAuth: false
        )

        // Store tokens and additional info
        self.accessToken = authResponse.accessToken
        self.refreshToken = authResponse.refreshToken
        self.tokenExpiresAt = Date().addingTimeInterval(TimeInterval(authResponse.expiresIn))
        self.deviceId = authResponse.deviceId
        self.userId = authResponse.userId
        self.organizationId = authResponse.organizationId

        // Store in keychain for persistence
        try await storeTokensInKeychain(authResponse)

        logger.info("Authentication successful")
        return authResponse
    }

    /// Get current access token, refreshing if needed
    public func getAccessToken() async throws -> String {
        // Check if token exists and is valid
        if let token = accessToken,
           let expiresAt = tokenExpiresAt,
           expiresAt > Date().addingTimeInterval(60) { // 1 minute buffer
            return token
        }

        // Try to refresh token if we have a refresh token
        if refreshToken != nil {
            return try await refreshAccessToken()
        }

        // Otherwise, we can't re-authenticate without API key
        throw RunAnywhereError.authenticationFailed("No valid token and no way to re-authenticate")
    }

    /// Perform health check
    public func healthCheck() async throws -> HealthCheckResponse {
        logger.debug("Performing health check")

        // Health check requires authentication
        return try await apiClient.get(
            .healthCheck,
            requiresAuth: true
        )
    }

    /// Check if authenticated
    public func isAuthenticated() -> Bool {
        return accessToken != nil
    }

    /// Clear authentication state
    public func clearAuthentication() async throws {
        accessToken = nil
        refreshToken = nil
        tokenExpiresAt = nil
        deviceId = nil
        userId = nil
        organizationId = nil

        // Clear from keychain
        try KeychainManager.shared.delete(for: "com.runanywhere.sdk.accessToken")
        try KeychainManager.shared.delete(for: "com.runanywhere.sdk.refreshToken")
        try KeychainManager.shared.delete(for: "com.runanywhere.sdk.deviceId")
        try KeychainManager.shared.delete(for: "com.runanywhere.sdk.userId")
        try KeychainManager.shared.delete(for: "com.runanywhere.sdk.organizationId")

        logger.info("Authentication cleared")
    }

    // MARK: - Private Methods

    private func refreshAccessToken() async throws -> String {
        guard let refreshToken = refreshToken else {
            throw RunAnywhereError.invalidAPIKey("No refresh token available")
        }

        guard let deviceId = deviceId else {
            throw RunAnywhereError.authenticationFailed("No device ID available for refresh")
        }

        logger.debug("Refreshing access token")

        let request = RefreshTokenRequest(
            deviceId: deviceId,
            refreshToken: refreshToken
        )

        // Call refresh endpoint
        let refreshResponse: RefreshTokenResponse = try await apiClient.post(
            .refreshToken,
            request,
            requiresAuth: false
        )

        // Update stored tokens and info
        self.accessToken = refreshResponse.accessToken
        self.refreshToken = refreshResponse.refreshToken
        self.tokenExpiresAt = Date().addingTimeInterval(TimeInterval(refreshResponse.expiresIn))
        self.deviceId = refreshResponse.deviceId
        self.userId = refreshResponse.userId
        self.organizationId = refreshResponse.organizationId

        // Store updated tokens in keychain
        try await storeTokensInKeychain(refreshResponse)

        logger.info("Token refresh successful")
        return refreshResponse.accessToken
    }

    private func storeTokensInKeychain(_ response: AuthenticationResponse) async throws {
        try KeychainManager.shared.store(response.accessToken, for: "com.runanywhere.sdk.accessToken")
        try KeychainManager.shared.store(response.refreshToken, for: "com.runanywhere.sdk.refreshToken")
        try KeychainManager.shared.store(response.deviceId, for: "com.runanywhere.sdk.deviceId")
        if let userId = response.userId {
            try KeychainManager.shared.store(userId, for: "com.runanywhere.sdk.userId")
        }
        try KeychainManager.shared.store(response.organizationId, for: "com.runanywhere.sdk.organizationId")
    }

    /// Load tokens from keychain if available
    public func loadStoredTokens() async throws {
        if let storedAccessToken = try? KeychainManager.shared.retrieve(for: "com.runanywhere.sdk.accessToken") {
            self.accessToken = storedAccessToken
            logger.debug("Loaded stored access token from keychain")
        }

        if let storedRefreshToken = try? KeychainManager.shared.retrieve(for: "com.runanywhere.sdk.refreshToken") {
            self.refreshToken = storedRefreshToken
            logger.debug("Loaded stored refresh token from keychain")
        }

        if let storedDeviceId = try? KeychainManager.shared.retrieve(for: "com.runanywhere.sdk.deviceId") {
            self.deviceId = storedDeviceId
            logger.debug("Loaded stored device ID from keychain")
        }

        if let storedUserId = try? KeychainManager.shared.retrieve(for: "com.runanywhere.sdk.userId") {
            self.userId = storedUserId
            logger.debug("Loaded stored user ID from keychain")
        }

        if let storedOrgId = try? KeychainManager.shared.retrieve(for: "com.runanywhere.sdk.organizationId") {
            self.organizationId = storedOrgId
            logger.debug("Loaded stored organization ID from keychain")
        }
    }

    /// Get current device ID
    public func getDeviceId() -> String? {
        return deviceId
    }

    /// Get current user ID
    public func getUserId() -> String? {
        return userId
    }

    /// Get current organization ID
    public func getOrganizationId() -> String? {
        return organizationId
    }

    /// Register device with backend
    /// Uses unified DeviceRegistrationRequest for all environments
    public func registerDevice() async throws -> DeviceRegistrationResponse {
        logger.debug("Registering device with backend")

        // Create request from current device info
        let request = DeviceRegistrationRequest.fromCurrentDevice()

        // Make API call with authentication
        let response: DeviceRegistrationResponse = try await apiClient.post(
            .deviceRegistration,
            request,
            requiresAuth: true
        )

        logger.info("Device registration successful: \(response.deviceId)")
        return response
    }
}

File: sdk/runanywhere-flutter/lib/data/network/services/authentication_service.dart
import 'dart:async';

import '../../../foundation/configuration/sdk_constants.dart';
import '../../../foundation/logging/sdk_logger.dart';
import '../../../foundation/security/keychain_manager.dart';
import '../../../infrastructure/device/services/device_identity.dart';
import '../../errors/repository_error.dart';
import '../api_client.dart';
import '../api_endpoint.dart';
import '../models/auth/auth.dart';

/// Service responsible for authentication and token management.
///
/// Matches iOS `AuthenticationService` actor from RunAnywhere SDK.
/// Implements [AuthTokenProvider] for use with [APIClient].
class AuthenticationService implements AuthTokenProvider {
  // MARK: - Properties

  final APIClient _apiClient;
  final SDKLogger _logger = SDKLogger(category: 'AuthenticationService');

  String? _accessToken;
  String? _refreshToken;
  DateTime? _tokenExpiresAt;
  String? _deviceId;
  String? _userId;
  String? _organizationId;

  // Keychain keys (matching iOS)
  static const _accessTokenKey = 'com.runanywhere.sdk.accessToken';
  static const _refreshTokenKey = 'com.runanywhere.sdk.refreshToken';
  static const _deviceIdKey = 'com.runanywhere.sdk.deviceId';
  static const _userIdKey = 'com.runanywhere.sdk.userId';
  static const _organizationIdKey = 'com.runanywhere.sdk.organizationId';

  // MARK: - Initialization

  AuthenticationService({required APIClient apiClient})
      : _apiClient = apiClient;

  /// Create and configure authentication services for production/staging.
  ///
  /// Returns a configured [AuthenticationService] that's already authenticated.
  /// Throws if authentication fails.
  static Future<AuthenticationService> createAndAuthenticate({
    required Uri baseURL,
    required String apiKey,
  }) async {
    final apiClient = APIClient(baseURL: baseURL, apiKey: apiKey);
    final authService = AuthenticationService(apiClient: apiClient);
    apiClient.setAuthTokenProvider(authService);

    // Authenticate with backend
    await authService.authenticate(apiKey: apiKey);

    return authService;
  }

  /// Get the associated API client.
  APIClient get apiClient => _apiClient;

  // MARK: - Public Methods

  /// Authenticate with the backend and obtain access token.
  Future<AuthenticationResponse> authenticate({required String apiKey}) async {
    final deviceId = await DeviceIdentity.persistentUUID;

    final request = AuthenticationRequest(
      apiKey: apiKey,
      deviceId: deviceId,
      platform: SDKConstants.platform,
      sdkVersion: SDKConstants.version,
    );

    _logger.debug('Authenticating with backend');

    // Use APIClient for the authentication request (doesn't require auth)
    final authResponse = await _apiClient.post<AuthenticationResponse>(
      APIEndpoint.authenticate,
      request,
      requiresAuth: false,
      fromJson: AuthenticationResponse.fromJson,
    );

    // Store tokens and additional info
    _accessToken = authResponse.accessToken;
    _refreshToken = authResponse.refreshToken;
    _tokenExpiresAt =
        DateTime.now().add(Duration(seconds: authResponse.expiresIn));
    _deviceId = authResponse.deviceId;
    _userId = authResponse.userId;
    _organizationId = authResponse.organizationId;

    // Store in keychain for persistence
    await _storeTokensInKeychain(authResponse);

    _logger.info('Authentication successful');
    return authResponse;
  }

  /// Get current access token, refreshing if needed.
  /// Implements [AuthTokenProvider.getAccessToken].
  @override
  Future<String> getAccessToken() async {
    // Check if token exists and is valid (with 1 minute buffer)
    if (_accessToken != null && _tokenExpiresAt != null) {
      final bufferTime = DateTime.now().add(const Duration(minutes: 1));
      if (_tokenExpiresAt!.isAfter(bufferTime)) {
        return _accessToken!;
      }
    }

    // Try to refresh token if we have a refresh token
    if (_refreshToken != null) {
      return await _refreshAccessToken();
    }

    // Otherwise, we can't re-authenticate without API key
    throw const RepositoryAuthenticationError(
        'No valid token and no way to re-authenticate');
  }

  /// Perform health check.
  Future<HealthCheckResponse> healthCheck() async {
    _logger.debug('Performing health check');

    // Health check requires authentication
    return await _apiClient.get<HealthCheckResponse>(
      APIEndpoint.healthCheck,
      requiresAuth: true,
      fromJson: HealthCheckResponse.fromJson,
    );
  }

  /// Check if authenticated.
  bool get isAuthenticated => _accessToken != null;

  /// Clear authentication state.
  Future<void> clearAuthentication() async {
    _accessToken = null;
    _refreshToken = null;
    _tokenExpiresAt = null;
    _deviceId = null;
    _userId = null;
    _organizationId = null;

    // Clear from keychain
    await KeychainManager.shared.delete(_accessTokenKey);
    await KeychainManager.shared.delete(_refreshTokenKey);
    await KeychainManager.shared.delete(_deviceIdKey);
    await KeychainManager.shared.delete(_userIdKey);
    await KeychainManager.shared.delete(_organizationIdKey);

    _logger.info('Authentication cleared');
  }

  /// Load tokens from keychain if available.
  Future<void> loadStoredTokens() async {
    final storedAccessToken =
        await KeychainManager.shared.retrieve(_accessTokenKey);
    if (storedAccessToken != null) {
      _accessToken = storedAccessToken;
      _logger.debug('Loaded stored access token from keychain');
    }

    final storedRefreshToken =
        await KeychainManager.shared.retrieve(_refreshTokenKey);
    if (storedRefreshToken != null) {
      _refreshToken = storedRefreshToken;
      _logger.debug('Loaded stored refresh token from keychain');
    }

    final storedDeviceId = await KeychainManager.shared.retrieve(_deviceIdKey);
    if (storedDeviceId != null) {
      _deviceId = storedDeviceId;
      _logger.debug('Loaded stored device ID from keychain');
    }

    final storedUserId = await KeychainManager.shared.retrieve(_userIdKey);
    if (storedUserId != null) {
      _userId = storedUserId;
      _logger.debug('Loaded stored user ID from keychain');
    }

    final storedOrgId =
        await KeychainManager.shared.retrieve(_organizationIdKey);
    if (storedOrgId != null) {
      _organizationId = storedOrgId;
      _logger.debug('Loaded stored organization ID from keychain');
    }
  }

  /// Get current device ID.
  String? get deviceId => _deviceId;

  /// Get current user ID.
  String? get userId => _userId;

  /// Get current organization ID.
  String? get organizationId => _organizationId;

  // MARK: - Private Methods

  Future<String> _refreshAccessToken() async {
    if (_refreshToken == null) {
      throw const RepositoryAuthenticationError('No refresh token available');
    }

    if (_deviceId == null) {
      throw const RepositoryAuthenticationError(
          'No device ID available for refresh');
    }

    _logger.debug('Refreshing access token');

    final request = RefreshTokenRequest(
      deviceId: _deviceId!,
      refreshToken: _refreshToken!,
    );

    // Call refresh endpoint
    final refreshResponse = await _apiClient.post<RefreshTokenResponse>(
      APIEndpoint.refreshToken,
      request,
      requiresAuth: false,
      fromJson: RefreshTokenResponse.fromJson,
    );

    // Update stored tokens and info
    _accessToken = refreshResponse.accessToken;
    _refreshToken = refreshResponse.refreshToken;
    _tokenExpiresAt =
        DateTime.now().add(Duration(seconds: refreshResponse.expiresIn));
    _deviceId = refreshResponse.deviceId;
    _userId = refreshResponse.userId;
    _organizationId = refreshResponse.organizationId;

    // Store updated tokens in keychain
    await _storeTokensInKeychain(refreshResponse);

    _logger.info('Token refresh successful');
    return refreshResponse.accessToken;
  }

  Future<void> _storeTokensInKeychain(AuthenticationResponse response) async {
    await KeychainManager.shared.store(_accessTokenKey, response.accessToken);
    await KeychainManager.shared.store(_refreshTokenKey, response.refreshToken);
    await KeychainManager.shared.store(_deviceIdKey, response.deviceId);
    if (response.userId != null) {
      await KeychainManager.shared.store(_userIdKey, response.userId!);
    }
    await KeychainManager.shared
        .store(_organizationIdKey, response.organizationId);
  }
}
